### PR TITLE
Add pretrain

### PR DESCRIPTION
**Pull request by nsorros**
The goal is to create a CLI tool and Python class to make it easier to pretrain models in the future. So for Kaggle and future clients we can use this library to easily pretrain models.

We care about masked language modelling (instead of causal language modelling) and also not about sentence prediction at this point but we can extend to those in the future.

I am implementing the training loop instead of using a Trainer to have more flexibility to tweak this to our needs. I am also thinking that we might want to pretrain general neural nets not only transformers.

### CODE REVIEW

**Comment by nsorros** [comment]
Would love some early thoughts @pdan93 @jehona-kryeziu

File: pretrain.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    def __init__(self, data, tokenizer, mask_percentage=0.15):
+        self.data = data
+        self.tokenizer = tokenizer
+        self.mask_percentage = mask_percentage
+
+    def __len__(self):
+        return len(self.data)
+
+    def __getitem__(self, idx):
+        inputs = self.tokenizer(self.data[idx], padding="max_length", truncation=True)
```
Comment (pdan93) [inline pretrain.py:31]: I think you need to specify the max_length param here as well

File: pretrain.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+@app.command()
+def pretrain(
+    data_path,
+    model_path,
+    pretrained_model="distilbert-base-uncased",
+    batch_size:int=32,
+    learning_rate:float=1e-5,
+    epochs:int=5,
+    mask_percentage:float=0.15,
+    dry_run:bool=False
```
Comment (pdan93) [inline pretrain.py:55]: love it! (the `dry_run` param)

File: pretrain.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        batches = tqdm(dataloader, desc=f"Epoch {epoch+1}/{epochs}")
+        for batch in batches:
+            inputs, labels = batch[0].to(device), batch[1].to(device)
+
+            optimizer.zero_grad()
+
+            outputs = model(inputs)
+            preds = torch.transpose(outputs.logits, 2, 1)
+
+            loss = criterion(preds, labels)
```
Comment (pdan93) [inline pretrain.py:91]: did this work? don't you have to do `labels.view(-1)` ?

**Review by pdan93 — APPROVED** [review]
Very nice! and yes, I think might be good to leave it possible to train normal NN's also. 
One little thing, would it make sense to make the loss function as a parameter? (with crossentropy as default). Don't think so, but asking...

One little idea, though it's not necessarily for this project. Wouldn't it be nice if we could somehow calculate the batch_size so that is uses as much memory as possible? (effectively meaning fastest process possible). And it would appear just as a suggestion

File: setup.cfg [hunk 4]
```diff
@@ -0,0 +1,28 @@
+[metadata]
+name = pretrain
+version = 0.0.1
```
Comment (ivyleavedtoadflax) [inline setup.cfg]: Consider starting with 0.1.0 to comply with [semver](https://semver.org/#how-should-i-deal-with-revisions-in-the-0yz-initial-development-phase)

**Comment by nsorros** [comment]
> Very nice! and yes, I think might be good to leave it possible to train normal NN's also. One little thing, would it make sense to make the loss function as a parameter? (with crossentropy as default). Don't think so, but asking...

Would it make sense to use another loss function? If so which?
 
> One little idea, though it's not necessarily for this project. Wouldn't it be nice if we could somehow calculate the batch_size so that is uses as much memory as possible? (effectively meaning fastest process possible). And it would appear just as a suggestion

This sounds interesting, let me think what's the best possible way to do that.

File: pretrain.py (hunk 3 above)
Comment (nsorros) [reply pretrain.py:91]: you could but it also accepts (N,d1) where N batch size and d1 sequence length in this case.

`labels.view(-1)` it collapses everything right? which is also acceptable

**Review by ivyleavedtoadflax — APPROVED** [review]
LGTM :rocket:

File: setup.cfg (hunk 4 above)
Comment (nsorros) [reply setup.cfg]: yep, good point, will change before this merges

File: pretrain.py (hunk 1 above)
Comment (nsorros) [reply pretrain.py:31]: I think it goes to the model max if not provided 
```
Pad to a maximum length specified with the argument max_length or to the maximum acceptable input length for the model if that argument is not provided.
```
from https://huggingface.co/docs/transformers/main/en/main_classes/tokenizer#transformers.PreTrainedTokenizer
